include/libopencm3/stm32/fdcan.h

```diff
@@ -0,0 +1,793 @@
+/**@}*/
+
+#define FDCAN_FIFO_EID_SHIFT			0
+#define FDCAN_FIFO_EID_MASK				(0x1FFFFFFF << FDCAN_FIFO_EID_SHIFT)
+
+#define FDCAN_FIFO_SID_SHIFT			18
+#define FDCAN_FIFO_SID_MASK				(0x7FF << FDCAN_FIFO_SID_SHIFT)
+
+#define FDCAN_FIFO_DLC_SHIFT			16
+#define FDCAN_FIFO_DLC_MASK				(0xF << FDCAN_FIFO_DLC_SHIFT)
+
+#define FDCAN_FIFO_MM_SHIFT				24
+#define FDCAN_FIFO_MM_MASK				(0xFF << FDCAN_FIFO_MM_SHIFT)
+
+#define FDCAN_FIFO_ANMF					(1 << 31)
+#define FDCAN_FIFO_FIDX_SHIFT			24
+#define FDCAN_FIFO_FIDX_MASK			(0x7F << FDCAN_FIFO_FIDX_SHIFT)
+
+#define FDCAN_FIFO_RXTS_SHIFT			0
+#define FDCAN_FIFO_RXTS_MASK			(0xFFFF << FDCAN_FIFO_RXTS_SHIFT)
+
+/** Message RAM layout for one FDCAN block.
+ * There are as many memory blocks as there are FDCAN blocks
+ */
+struct fdcan_message_ram {
+	/* List of standard ID filters */
+	struct fdcan_standard_filter lfssa[FDCAN_SFT_MAX_NR];
+
+	/* List of extended ID filters */
+	struct fdcan_extended_filter lfesa[FDCAN_EFT_MAX_NR];
+
+	/* Buffer area for two receive FIFOs each having space for three messages */
+	struct fdcan_rx_fifo_element rx_fifo[2][3];
+
+	/* Buffer area for transmit event buffers */
+	struct fdcan_tx_event_element tx_event[3];
+
+	/* Buffer area for transmitted messages. May act either as FIFO or as queue
+	 * depending on configuration
+	 */
+	struct fdcan_tx_buffer_element tx_buffer[3];
+};
+
+
+/* --- FD-CAN error returns ------------------------------------------------- */
+
+/** FDCAN error return values
+ */
+enum fdcan_error {
+	/** No error. Operation finished successfully */
+	FDCAN_E_OK,
+
+	/** Value provided was out of range */
+	FDCAN_E_OUTOFRANGE,
+
+	/** Timeout waiting for FDCAN block to accept INIT bit change */
+	FDCAN_E_TIMEOUT,
+
+	/** Value provided was invalid (FIFO index, FDCAN block base address, length, etc.) */
+	FDCAN_E_INVALID,
+
+	/** Device is busy: Transmit buffer is full, unable to queue additional message or device
+	 * is outside of INIT mode and cannot perform desired operation. */
+	FDCAN_E_BUSY,
+
+	/** Receive buffer is empty, unable to read any new message */
+	FDCAN_E_NOTAVAIL
+};
+
+/**@}*/
+
+/* --- FD-CAN functions ----------------------------------------------------- */
+
+BEGIN_DECLS
+
+int fdcan_init(uint32_t canport, uint32_t timeout);
+
+void fdcan_set_can(uint32_t canport, bool auto_retry_disable, bool rx_fifo_locked,
+		bool tx_queue_mode,	bool silent, uint32_t n_sjw, uint32_t n_ts1, uint32_t n_ts2,
+		uint32_t n_br_presc);
+
+void fdcan_set_fdcan(uint32_t canport, bool brs_enable, bool fd_op_enable,
+		uint32_t f_sjw, uint32_t f_ts1, uint32_t f_ts2, uint32_t f_br_presc);
+
+void fdcan_set_test(uint32_t canport, bool testing, bool loopback);
+
+void fdcan_filter_init(uint32_t canport, uint8_t std_filt, uint8_t ext_filt);
+
+int fdcan_start(uint32_t canport, uint32_t timeout);
+
+void fdcan_set_std_filter(uint32_t canport, uint32_t nr,
+		uint8_t id_list_mode, uint32_t id1, uint32_t id2,
+		uint8_t action);
```

Comment:
unless you _need_ this to be 8 bit, just use "int" here, so the compiler isn't being forced to mask things down to 8 bits for no reason.  (in general, this applies to all api places.  just "int" and "unsigned" will do the best thing for you normally, though you can go crazy with uint_fast8_t if you want to go nuts.... (I'd not really recommend using the "special" types in public apis though honestly.)

Reply:
(you know what, don't worry about this, I just went and played on godbolt for a while, and this advice is outdated these days, just leave this as is)